syntax = "proto2";

package Battlenet.JSON;

option optimize_for = CODE_SIZE;

enum FormType {
  LOGIN_FORM = 1;
}

message FormInput {
  required string input_id = 1;
  required string type = 2;
  required string label = 3;
  optional uint32 max_length = 4;
}

message FormInputs {
  required .Battlenet.JSON.FormType type = 1;
  repeated FormInput inputs = 2;
}

message FormInputValue {
  required string input_id = 1;
  required string value = 2;
}

message LoginForm {
  required string platform_id = 1;
  required string program_id = 2;
  required string version = 3;
  repeated .Battlenet.JSON.FormInputValue inputs = 4;
}

enum AuthenticationState {
  LOGIN = 1;
  LEGAL = 2;
  AUTHENTICATOR = 3;
  DONE = 4;
}

message LoginResult {
  required .Battlenet.JSON.AuthenticationState authentication_state = 1;
  optional string error_code = 2;
  optional string error_message = 3;
  optional string url = 4;
  optional string login_ticket = 5;
}

message RealmListTicketIdentity {
  required fixed32 gameAccountID = 1;
  required uint32 gameAccountRegion = 2;
}

message ClientVersion {
  required uint32 versionMajor = 1;
  required uint32 versionMinor = 2;
  required uint32 versionRevision = 3;
  required fixed32 versionBuild = 4;
}

message ClientInformation {
  required fixed32 platform = 1;
  required string buildVariant = 2;
  required fixed32 type = 3;
  required string timeZone = 4;
  required uint32 currentTime = 5;
  required fixed32 textLocale = 6;
  required fixed32 audioLocale = 7;
  required fixed32 versionDataBuild = 8;
  required .Battlenet.JSON.ClientVersion version = 9;
  required bytes secret = 10;
}

message RealmListTicketClientInformation {
  required .Battlenet.JSON.ClientInformation info = 1;
}

message RealmCharacterCountEntry {
  required fixed32 wowRealmAddress = 1;
  required uint32 count = 2;
}

message RealmCharacterCountList {
  repeated .Battlenet.JSON.RealmCharacterCountEntry counts = 1;
}

message RealmUpdate {
  required fixed32 wowRealmAddress = 1;
  required uint32 cfgTimezonesID = 2;
  required uint32 populationState = 3;
  required uint32 cfgCategoriesID = 4;
  required .Battlenet.JSON.ClientVersion version = 5;
  required uint32 cfgRealmsID = 6;
  required uint32 flags = 7;
  required string name = 8;
  required uint32 cfgConfigsID = 9;
  required uint32 cfgLanguagesID = 10;
}

message RealmState {
  optional .Battlenet.JSON.RealmUpdate update = 1;
  required bool deleting = 2;
}

message RealmListUpdates {
  repeated .Battlenet.JSON.RealmState updates = 1;
}

message IPAddress {
  required string ip = 1;
  required uint32 port = 2;
}

message RealmIPAddressFamily {
  required uint32 family = 1;
  repeated .Battlenet.JSON.IPAddress addresses = 2;
}

message RealmListServerIPAddresses {
  repeated .Battlenet.JSON.RealmIPAddressFamily families = 1;
}
